### PR TITLE
Fix `bound_claims` not getting unset for `vault_jwt_auth_backend_role`

### PR DESCRIPTION
**Pull request by lawliet89**
### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" comments, they generate extra noise for pull request followers and do not help prioritize the request

When `bound_claims` in `vault_jwt_auth_backend_role` is unset, it does not get removed from the role, resulting in a permanent diff and unexpected results.

Release note for [CHANGELOG](https://github.com/hashicorp/terraform-provider-vault/blob/master/CHANGELOG.md):

```release-note
Fix `bound_claims` not getting removed for `vault_jwt_auth_backend_role`
```

Output from acceptance testing:

```
$ $ VAULT_ADDR=http://127.0.0.1:8200 VAULT_TOKEN=12345 TESTARGS="--run TestAccJWTAuthBackendRole" make testacc
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test -v --run TestAccJWTAuthBackendRole -timeout 30m ./...
?       github.com/hashicorp/terraform-provider-vault   [no test files]
?       github.com/hashicorp/terraform-provider-vault/cmd/coverage      [no test files]
?       github.com/hashicorp/terraform-provider-vault/cmd/generate      [no test files]
testing: warning: no tests to run
PASS
ok      github.com/hashicorp/terraform-provider-vault/codegen   (cached) [no tests to run]
?       github.com/hashicorp/terraform-provider-vault/generated [no test files]
testing: warning: no tests to run
PASS
ok      github.com/hashicorp/terraform-provider-vault/generated/datasources/transform/decode    (cached) [no tests to run]
testing: warning: no tests to run
PASS
ok      github.com/hashicorp/terraform-provider-vault/generated/datasources/transform/encode    (cached) [no tests to run]
testing: warning: no tests to run
PASS
ok      github.com/hashicorp/terraform-provider-vault/generated/resources/transform/alphabet    (cached) [no tests to run]
testing: warning: no tests to run
PASS
ok      github.com/hashicorp/terraform-provider-vault/generated/resources/transform/role        (cached) [no tests to run]
testing: warning: no tests to run
PASS
ok      github.com/hashicorp/terraform-provider-vault/generated/resources/transform/template    (cached) [no tests to run]
testing: warning: no tests to run
PASS
ok      github.com/hashicorp/terraform-provider-vault/generated/resources/transform/transformation      (cached) [no tests to run]
?       github.com/hashicorp/terraform-provider-vault/helper    [no test files]
testing: warning: no tests to run
PASS
ok      github.com/hashicorp/terraform-provider-vault/internal/identity/entity  (cached) [no tests to run]
?       github.com/hashicorp/terraform-provider-vault/schema    [no test files]
?       github.com/hashicorp/terraform-provider-vault/testutil  [no test files]
testing: warning: no tests to run
PASS
ok      github.com/hashicorp/terraform-provider-vault/util      (cached) [no tests to run]
=== RUN   TestAccJWTAuthBackendRole_import
--- PASS: TestAccJWTAuthBackendRole_import (0.66s)
=== RUN   TestAccJWTAuthBackendRole_basic
--- PASS: TestAccJWTAuthBackendRole_basic (0.55s)
=== RUN   TestAccJWTAuthBackendRole_update
--- PASS: TestAccJWTAuthBackendRole_update (0.96s)
=== RUN   TestAccJWTAuthBackendRole_full
--- PASS: TestAccJWTAuthBackendRole_full (0.70s)
=== RUN   TestAccJWTAuthBackendRoleOIDC_full
--- PASS: TestAccJWTAuthBackendRoleOIDC_full (1.19s)
=== RUN   TestAccJWTAuthBackendRoleOIDC_disableParsing
--- PASS: TestAccJWTAuthBackendRoleOIDC_disableParsing (1.19s)
=== RUN   TestAccJWTAuthBackendRole_fullUpdate
--- PASS: TestAccJWTAuthBackendRole_fullUpdate (1.77s)
PASS
ok      github.com/hashicorp/terraform-provider-vault/vault     7.042s
```

If the `else` clause I added in the resource is removed, the test above fails with

```
=== RUN   TestAccJWTAuthBackendRole_fullUpdate
    resource_jwt_auth_backend_role_test.go:371: Step 3/3 error: Check failed: 1 error occurred:
                * Check 23/23 error: vault_jwt_auth_backend_role.role: Attribute 'bound_claims.%' expected "0", got "2"
```

thereby confirming the bug.

### CODE REVIEW

File: vault/resource_jwt_auth_backend_role.go [hunk 1]
```diff
@@ -406,6 +406,8 @@ func jwtAuthBackendRoleDataToWrite(d *schema.ResourceData, create bool) map[stri
 			boundClaims[key] = claims
 		}
 		data["bound_claims"] = boundClaims
+	} else {
+		data["bound_claims"] = make(map[string]interface{}, 0)
```
Comment (benashz) [inline vault/resource_jwt_auth_backend_role.go:410]: The empty map is defined on line [396](https://github.com/hashicorp/terraform-provider-vault/pull/1469/files#diff-4ce5833472afbb5d5a963d14f398fcba1ff7736c48af6b03f49cf3bca0824537R396) above. If we move that before the conditional we can drop the `else{}` and always set `bound_claims`

File: vault/resource_jwt_auth_backend_role_test.go [hunk 2]
```diff
@@ -477,6 +479,58 @@ func TestAccJWTAuthBackendRole_fullUpdate(t *testing.T) {
 						"verbose_oidc_logging", "false"),
 				),
 			},
+			// Repeat test case again to remove attributes like `bound_claims`
```
Comment (benashz) [inline vault/resource_jwt_auth_backend_role_test.go:482]: Would you mind moving these test check funcs to a common set that can be shared between the various steps. Please use https://github.com/hashicorp/terraform-provider-vault/blob/e3b6f4602c74e7f3a3e9f1f96e4f031453c54792/vault/resource_pki_secret_backend_role_test.go#L72 as an example.

The resource name/path should also be factored out to a test variable called `resourceName`, e.g. https://github.com/hashicorp/terraform-provider-vault/blob/e3b6f4602c74e7f3a3e9f1f96e4f031453c54792/vault/resource_pki_secret_backend_role_test.go#L18

Thanks

File: vault/resource_jwt_auth_backend_role.go (hunk 1 above)
Comment (lawliet89) [reply vault/resource_jwt_auth_backend_role.go]: I've done this.

File: vault/resource_jwt_auth_backend_role_test.go (hunk 2 above)
Comment (lawliet89) [reply vault/resource_jwt_auth_backend_role_test.go:484]: I reduced a bit of code reptition but I did not refactor out the test steps into a separate function yet. This seems like more of a refactoring that might make the size of this PR bigger.

Comment (benashz) [reply vault/resource_jwt_auth_backend_role_test.go:484]: Ok, we will take care of that after merge. Thanks!

**Review by benashz — APPROVED** [review]
Thank for your contribution to HashiCorp!

**Pull request merged**